Test suite

nickhrdy opened the pull request:
This is the same test suite that was mentioned in #1.  The suite currently uses 4 topologies for testing with 12 configurations tested on each. The README in the test folder has information on the test and the format of the answers files. 

Closes #3

nickhrdy committed:
tests: Add test suite

nickhrdy committed:
tests: Add tests for environment variables

nickhrdy committed:
tests_utils: Update testing to handle smt-1 topos

nickhrdy committed:
Update tests and add new test

eleon commented on tests/Makefile.am:
```diff
@@ -0,0 +1,26 @@
+AM_CFLAGS= -Wall -Werror -I$(top_srcdir)/src $(HWLOC_CFLAGS)
+LDADD = ../src/libmpibind.la libtap.la $(HWLOC_LIBS)
```
```suggestion
LDADD = $(top_srcdir)/src/libmpibind.la libtap.la $(HWLOC_LIBS)
```

eleon commented on tests/Makefile.am:
```diff
@@ -0,0 +1,26 @@
+
+check_PROGRAMS = $(TESTS)
+
+TEST_EXTENSIONS = .t
```
How is this variable used? Is there anything special about its name?

nickhrdy replied:
anything listed under check_PROGRAMS is only built when running make check. The same thing can apply to other variables like check_LIBRARIES and check_LTLIBRARIES, etc.

The TEST_EXTENSIONS variable modifies how the log files are named. [Here's a link that explains some of the variables we can use to configure the test harness.](https://www.gnu.org/software/automake/manual/html_node/Parallel-Test-Harness.html)

nickhrdy committed:
tests: replace relative path with absolute path

Co-authored-by: Edgar Leon <800736+eleon@users.noreply.github.com>

nickhrdy committed:
tests: update build to check HAVE_LIBTAP variable

nickhrdy committed:
tests: conform to provided skeleton

nickhrdy committed:
Remove tap-driver.sh from repo
Now it's automatically added suring the bootstrap

nickhrdy committed:
Remove libtap source file
We will depend on the user to install them separately

eleon commented on config/ar-lib:
```diff
@@ -0,0 +1,270 @@
+#! /bin/sh
+# Wrapper for Microsoft lib.exe
```
Where does this file come from (ar-lib)? Can it be autogenerated by autotools?

nickhrdy replied:
This is my bad. it's generated by the autotools and I thought I had removed it from the repo already

eleon commented on test-suite/test_utils.c:
```diff
@@ -0,0 +1,490 @@
+    return 1;
+  }
+
+  hdl->ntasks = 0;
```
An mpibind handle is an opaque object. It should not be used directly. The structure behind it may change and should not break the test suite. 

```suggestion
  hdl->ntasks = 0;
```

nickhrdy replied:
I replaced the described cases with the new getters. One thing to note is that I did make an almost identical struct called mpibind_test_t. The reason for this was so that I wouldn't be parsing test answers straight into the mpibind_t handles, and so that if needed, we could add members to the imitation struct to hold information that might be useful for current or future tests.

eleon commented on test-suite/test_utils.h:
```diff
@@ -0,0 +1,81 @@
+#include <string.h>
+
+#include "mpibind.h"
+#include "mpibind-priv.h"
```
mpibind-priv.h should not be used by user programs. The only header for users/tests is 'mpibind.h'

```suggestion
#include "mpibind-priv.h"
```

nickhrdy replied:
This has been removed from test_utils.h

nickhrdy committed:
Remove config/ar-lib from repo

nickhrdy committed:
Add getters for input parameters (#12)

in_nthreads didn't get one as not to confuse it with nthreads

eleon committed:
Reordering new getter functions and renaming a couple to match their corresponding setter functions.

nickhrdy committed:
tests: Update utils to use getters

nickhrdy committed:
tests: add error checks for getters

nickhrdy committed:
Update test struct names and how the quantity of tests and answers are checked
Now the number of tests is specified in the .h file instead of comparing across different files
